### PR TITLE
Nested Loops

### PR DESCRIPTION
**Pull request by yashmit-alshaya**
Screenshot 2023-09-18 at 3 26 36 PM

### CODE REVIEW

File: src/Context/GtmContext.php [hunk 1]
```diff
@@ -75,17 +75,46 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
-    foreach ($json_arr as $json_item) {
-      if (isset($json_item[$key])) {
-        return $json_item[$key];
-      }
+    protected function getDataLayerValue($key) {
+        $json_arr = $this->getDataLayerJson();
+        // Loop through the array and return the data layer value
+        foreach ($json_arr as $json_item) {
+            // Check if the key contains dot.
+            if (strpos($key, '.', 0) !== false) {
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: @yashmit-alshaya Lets move this entire block into a new function, so we keep getDataLayerValue() small.
I suggest the name `getDotValue()`
In the function annotation, please explain the usage with examples.

File: src/Context/GtmContext.php [hunk 2]
```diff
@@ -75,17 +75,46 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+                if (strpos($key, '[', 0) !== false) {
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Please add a comment to explain this if condition

File: src/Context/GtmContext.php [hunk 3]
```diff
@@ -75,17 +75,46 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+                    // Trim before [
+                    $pos = strpos($key, '[');
+                    if ($pos !== false) {
+                        $trimmedString = substr($key, 0, $pos);
+                    }
+                    $value = $dot->get($trimmedString);
+                    if ($value !== NULL) {
+                        // Filter the Numbers from String
+                        $int_var = (int)filter_var($key, FILTER_SANITIZE_NUMBER_INT);
+                        //Filter last variable from key
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Missing space after //

File: src/Context/GtmContext.php [hunk 4]
```diff
@@ -75,17 +75,46 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+                if (strpos($key, '[', 0) !== false) {
+                    $dot = dot($json_item);
+                    // Trim before [
+                    $pos = strpos($key, '[');
+                    if ($pos !== false) {
+                        $trimmedString = substr($key, 0, $pos);
+                    }
+                    $value = $dot->get($trimmedString);
+                    if ($value !== NULL) {
+                        // Filter the Numbers from String
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Comments should end with a .

File: src/Context/GtmContext.php [hunk 5]
```diff
@@ -75,17 +75,46 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+                        $int_var = (int)filter_var($key, FILTER_SANITIZE_NUMBER_INT);
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: missing space after )

File: src/Context/GtmContext.php [hunk 6]
```diff
@@ -75,17 +75,46 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
-        return $json_item[$key];
-      }
+    protected function getDataLayerValue($key) {
+        $json_arr = $this->getDataLayerJson();
+        // Loop through the array and return the data layer value
+        foreach ($json_arr as $json_item) {
+            // Check if the key contains dot.
+            if (strpos($key, '.', 0) !== false) {
+                if (strpos($key, '[', 0) !== false) {
+                    $dot = dot($json_item);
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: This line is present on the if and the else condition, can we move it before the if condition?

File: src/Context/GtmContext.php [hunk 7]
```diff
@@ -75,17 +75,46 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+                    }
+                    $value = $dot->get($trimmedString);
+                    if ($value !== NULL) {
+                        // Filter the Numbers from String
+                        $int_var = (int)filter_var($key, FILTER_SANITIZE_NUMBER_INT);
+                        //Filter last variable from key
+                        $lastDotPosition = strrpos($key, ".");
+                        if ($lastDotPosition !== false) {
+                            $last_variable = substr($key, $lastDotPosition + 1);
+                            print_r(__LINE__, $last_variable);
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Why this print_r here?

File: src/Context/GtmContext.php (hunk 1 above)
Comment (yashmit-alshaya) [reply src/Context/GtmContext.php]: Yes created a new function and changed other as per suggestion.

File: src/Context/GtmContext.php [hunk 8]
```diff
@@ -75,17 +76,58 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
 
-    // Loop through the array and return the data layer value
-    foreach ($json_arr as $json_item) {
-      if (isset($json_item[$key])) {
-        return $json_item[$key];
-      }
+    /**
+     * Convert dataLayer values to dot notation
+     */
+    private function getDotValue($arr, $key)
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Please add annotations to the arguments and return values

File: src/Context/GtmContext.php [hunk 9]
```diff
@@ -75,17 +76,58 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+        throw new \Exception($key . ' not found.');
+    }
 
-    // Loop through the array and return the data layer value
-    foreach ($json_arr as $json_item) {
-      if (isset($json_item[$key])) {
-        return $json_item[$key];
-      }
+    /**
+     * Convert dataLayer values to dot notation
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Converts arrays to dot notation.

File: src/Context/GtmContext.php [hunk 10]
```diff
@@ -75,17 +76,58 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
    * @return mixed
    * @throws \Exception
    */
-  protected function getDataLayerValue($key) {
-    $json_arr = $this->getDataLayerJson();
+    protected function getDataLayerValue($key)
+    {
+        $json_arr = $this->getDataLayerJson();
+        // Loop through the array and return the data layer value
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Missing . at the end of line

File: src/Context/GtmContext.php [hunk 11]
```diff
@@ -75,17 +76,58 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+        foreach ($json_arr as $json_item) {
+            // Check if the key contains dot.
+            if (strpos($key, '.', 0) !== false) {
+                $value = $this->getDotValue($json_item, $key);
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Add a comment here: `// Get value using dot notation.`

File: src/Context/GtmContext.php (hunk 8 above)
Comment (marcelovani) [reply src/Context/GtmContext.php]: Please rename this variable from `$arr` to `$value`

File: src/Context/GtmContext.php [hunk 12]
```diff
@@ -75,17 +76,58 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
-      if (isset($json_item[$key])) {
-        return $json_item[$key];
-      }
+    /**
+     * Convert dataLayer values to dot notation
+     */
+    private function getDotValue($arr, $key)
+    {
+        $dot = dot($arr);
+        // Check if the array contains dot.
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: This comment is not accurate. It should say
```
// Check if the key contains [.
// When the key contains [0] it means that we are trying to get the value from specific index in an array
// i.e. foo.bar[0].foo
```

We also need to add a comment in the main Readme to explain how to use this into a Feature file.

File: src/Context/GtmContext.php [hunk 13]
```diff
@@ -75,17 +76,58 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+        if (strpos($key, '[', 0) !== false) {
+            // Trim before [
+            $pos = strpos($key, '[');
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: This is the same code you have on line 104.
Please move this to line 103 and use `$pos` in the comparison on line 104

File: src/Context/GtmContext.php [hunk 14]
```diff
@@ -75,17 +76,58 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+            if ($pos !== false) {
+                $trimmedString = substr($key, 0, $pos);
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: This is not really trimmed string, this is actually the key before the index, please update the variable and add a comment to explain this.

File: src/Context/GtmContext.php [hunk 15]
```diff
@@ -75,17 +76,58 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+            }
+            $value = $dot->get($trimmedString);
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: I believe this variable will contain the array of values. Please rename `$value` to `$values`

File: src/Context/GtmContext.php [hunk 16]
```diff
@@ -75,17 +76,58 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+            if ($value !== NULL) {
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Since we expect an array, we should use `if (!empty(...`

File: src/Context/GtmContext.php [hunk 17]
```diff
@@ -75,17 +76,58 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+                // Filter the number from string.
+                $int_var = (int)filter_var($key, FILTER_SANITIZE_NUMBER_INT);
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: This is bad code. If the keys contains numbers, it will cause problems.
Please use regex to find the key by extracting numbers between the brackets `[1]`

File: src/Context/GtmContext.php [hunk 18]
```diff
@@ -75,17 +76,58 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+                // Filter last variable from key.
+                $lastDotPosition = strrpos($key, ".");
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: This is bad code, it would not work with a key like `products.list[11].item.group.name`
Please use strpos to get the value that come after the closing bracket, in this case it should result in `item.group.name`

File: src/Context/GtmContext.php (hunk 11 above)
Comment (yashmit-alshaya) [reply src/Context/GtmContext.php]: Done

File: src/Context/GtmContext.php [hunk 19]
```diff
@@ -75,17 +76,69 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
    * @return mixed
    * @throws \Exception
    */
-  protected function getDataLayerValue($key) {
-    $json_arr = $this->getDataLayerJson();
+    protected function getDataLayerValue($key)
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Please revert changes on this line and below

File: src/Context/GtmContext.php [hunk 20]
```diff
@@ -75,17 +76,69 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
 
-    // Loop through the array and return the data layer value
-    foreach ($json_arr as $json_item) {
-      if (isset($json_item[$key])) {
-        return $json_item[$key];
-      }
+    /**
+     * Convert arrays :arg1 to dot notation :arg2
+     */
+    private function getDotValue($value, $key)
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Please add proper PHPDoc with @param and @return tags. see https://docs.devsense.com/en/vs/editor/phpdoc#:~:text=There%20are%20several%20PHPDoc%20tags,(%24name)%20(description)%20.

File: src/Context/GtmContext.php [hunk 21]
```diff
@@ -75,17 +76,69 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+            }
+        }
     }
-    throw new \Exception($key . ' not found.');
-  }
+
+  /**
+   * Get number inside array index :arg2
+   */
+   protected function getNumberInsideKey($key) {
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Please add proper PHPDoc with @param and @return tags

File: src/Context/GtmContext.php [hunk 22]
```diff
@@ -75,17 +76,69 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+     */
+    private function getDotValue($value, $key)
+    {
+        $dot = dot($value);
+        // Check if the key contains [.
+        // When the key contains [0] it means that we are trying to get the value from specific index in an array.
+        // i.e. foo.bar[0].foo.
+        $pos = strpos($key, '[',0);
+        if ($pos) {
+            // Trim before [
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Missing .

File: src/Context/GtmContext.php [hunk 23]
```diff
@@ -75,17 +76,69 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+            $keyBeforeArray = substr($key, 0, $pos);
+            $values = $dot->get($keyBeforeArray);
+            if (!empty($values)) {
+                // Filter the number from string.
+                $int_var = $this->getNumberInsideKey($key);
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Please do not mix Camel Case and Snake Case, all variables should follow same pattern. I would recommend you to rename this variable to $index.

File: src/Context/GtmContext.php (hunk 12 above)
Comment (marcelovani) [reply src/Context/GtmContext.php]: Please update Readme with this new feature https://github.com/marcelovani/behat-gtm/blob/54c1b9f5ebd29c946afb1e1c8722ed7aae90ec33/README.md

File: src/Context/GtmContext.php (hunk 21 above)
Comment (marcelovani) [inline src/Context/GtmContext.php]: Please rename this function to `getNumberFromString`

File: src/Context/GtmContext.php [hunk 24]
```diff
@@ -75,17 +76,69 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+                // Filter last variable from key.
+                $arrayIndexPosition = strrpos($key, "].");
+                if ($arrayIndexPosition) {
+                    $keyAfterArray = substr($key, $arrayIndexPosition + 2);
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Pls rename $keyAfterArray to $key

File: src/Context/GtmContext.php [hunk 25]
```diff
@@ -75,17 +76,69 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+                return $dot_array->get($keyAfterArray);
+            }
+        } else {
+            $dot = dot($value);
+            $valueNonArray = $dot->get($key);
+            if ($valueNonArray !== NULL) {
+                return $valueNonArray;
+            }
+        }
     }
```
Comment (marcelovani) [inline src/Context/GtmContext.php:125]: This code can be refactored to something like:

```                }
                $value = dot($values[$int_var]);
            }
        }
        $dot = dot($value);
        $value = $dot->get($key);
        if (!is_null($value)) {
            return $value;
        }
```

File: src/Context/GtmContext.php [hunk 26]
```diff
@@ -77,16 +78,70 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
   }
 
+  /**
+   * Convert arrays to dot notation
+   * @param $value
+   * @param $key
+   * @return mixed|void
+   */
+  private function getDotValue($value, $key) {
+    $dot = dot($value);
```
Comment (marcelovani) [inline src/Context/GtmContext.php:106]: Please delete this line

File: src/Context/GtmContext.php [hunk 27]
```diff
@@ -77,16 +78,70 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+        }
+      } elseif (isset($json_item[$key])) {
+          return $json_item[$key];
       }
     }
     throw new \Exception($key . ' not found.');
   }
 
+  /**
+   * Convert arrays to dot notation
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Please add . at the end of sentence and add a blank line after this line.

File: src/Context/GtmContext.php (hunk 25 above)
Comment (marcelovani) [reply src/Context/GtmContext.php:125]: @yashmit-alshaya Actually we can make this simply:

```
$dot = dot($value);
return $dot->get($key);
```

Reason: we don't need a new variable and we dont need to check for is_null.

File: src/Context/GtmContext.php [hunk 28]
```diff
@@ -77,16 +78,70 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+   * @param $value
+   * @param $key
+   * @return mixed|void
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: @return mixed|null

File: src/Context/GtmContext.php [hunk 29]
```diff
@@ -77,16 +78,70 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+    }
+    $dot = dot($value);
+    $value = $dot->get($key);
+    if (!is_null($value)) {
+      return $value;
+    }
+  }
+
+  /**
+   * Get number inside array index
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Pls add . at the end of line and a new line after this line

File: src/Context/GtmContext.php [hunk 30]
```diff
@@ -77,16 +78,70 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+   * @param $key
+   * @return string
+   * @throws \Exception
+   */
+   protected function getNumberFromString($key) {
+     $pattern = '/\[(\d+)]/';
+     if (preg_match_all($pattern, $key, $matches)) {
+       return $matches[1][0];
+     }
+     throw new \Exception('Number not found between' . $key );
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: `throw new \Exception('Number not found in key ' . $key );`

File: src/Context/GtmContext.php (hunk 26 above)
Comment (yashmit-alshaya) [reply src/Context/GtmContext.php:106]: Can't delete this because in next line we are using this dot function 
`$values = $dot->get($keyBeforeArray);`

File: src/Context/GtmContext.php [hunk 31]
```diff
@@ -77,16 +78,69 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
    */
   protected function getDataLayerValue($key) {
     $json_arr = $this->getDataLayerJson();
-
-    // Loop through the array and return the data layer value
+    // Loop through the array and return the data layer value.
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Pls undo this change

File: README.md [hunk 32]
```diff
@@ -21,3 +21,37 @@ Given google tag manager data layer setting :key should match :value
...
+            "position": "1"
+         },
+      ]
+   },
+   "gtm.uniqueEventId": 127
+}
+```
+To extract the GTM value nested in arrays i.e name under impressions. We can use dot notation
+```
+ecommerce.impressions[0].name
```
Comment (marcelovani) [inline README.md]: `Given google tag manager data layer setting "ecommerce.impressions[0].name" should match "~Knockout~"`

File: src/Context/GtmContext.php (hunk 26 above)
Comment (marcelovani) [reply src/Context/GtmContext.php:106]: Ok, then I will suggest another change below on line 122

File: src/Context/GtmContext.php [hunk 33]
```diff
@@ -77,16 +78,69 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+        $index = $this->getNumberFromString($key);
+        // Filter last variable from key.
+        $arrayIndexPosition = strrpos($key, "].");
+        if ($arrayIndexPosition) {
+          $key = substr($key, $arrayIndexPosition + 2);
+        }
+        $value = dot($values[$index]);
+      }
+    }
+    $dot = dot($value);
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Pls delete this line as it is the same as line 105

File: src/Context/GtmContext.php [hunk 34]
```diff
@@ -77,16 +78,69 @@ public function getDataLayerSettingShouldMatch($key, $regex) {
...
+      $values = $dot->get($keyBeforeArray);
+      if (!empty($values)) {
+        // Filter the number from string.
+        $index = $this->getNumberFromString($key);
+        // Filter last variable from key.
+        $arrayIndexPosition = strrpos($key, "].");
+        if ($arrayIndexPosition) {
+          $key = substr($key, $arrayIndexPosition + 2);
+        }
+        $value = dot($values[$index]);
```
Comment (marcelovani) [inline src/Context/GtmContext.php]: Pls change this line to `$value = dot($values[$index]);`
